blockscout-ens/bens-logic/src/protocols/protocoler.rs

```diff
@@ -288,17 +290,114 @@ impl Protocoler {
+                        network_id,
+                        maybe_filter.clone(),
+                    )
+                    .unwrap_or_default()
+                })
+                .take(MAX_NAMES_LIMIT)
+                .collect();
+
+            if all_names_with_protocols.is_empty() {
+                Err(ProtocolError::InvalidName {
+                    name: name.to_string(),
+                    reason: "No valid TLDs".to_string(),
+                })
+            } else {
+                Ok(all_names_with_protocols)
+            }
+        }
+    }
+
+    pub fn names_options_in_network_exact(
+        &self,
+        name: &str,
+        network_id: i64,
+        maybe_filter: Option<NonEmpty<String>>,
     ) -> Result<Vec<DomainNameOnProtocol>, ProtocolError> {
         let tld = Tld::from_domain_name(name).ok_or_else(|| ProtocolError::InvalidName {
             name: name.to_string(),
-            reason: "no tld found".to_string(),
+            reason: "Invalid TLD".to_string(),
         })?;
+
         let protocols = self.protocols_of_network_for_tld(network_id, tld, maybe_filter)?;
-        let names_with_protocols = protocols
-            .into_iter()
-            .map(|p| DomainNameOnProtocol::from_str(name, p))
-            .collect::<Result<_, _>>()?;
-        Ok(names_with_protocols)
+
+        let mut results = Vec::new();
+        for deployed_protocol in protocols {
+            let empty_label_hash = match &deployed_protocol.protocol.info.protocol_specific {
+                ProtocolSpecific::EnsLike(ens_like) => ens_like.empty_label_hash,
+                ProtocolSpecific::D3Connect(d3_connect) => d3_connect.empty_label_hash,
+            };
+
+            let domain_name = DomainName::new(name, empty_label_hash)?;
+            results.push(DomainNameOnProtocol {
+                inner: domain_name,
+                deployed_protocol,
+            });
+        }
+
+        Ok(results)
+    }
+
+    fn names_options_in_network_with_suggestions(
+        &self,
+        name_with_tld: &str,
+        network_id: i64,
+        maybe_filter: Option<NonEmpty<String>>,
+    ) -> Result<Vec<DomainNameOnProtocol>, ProtocolError> {
+        let protocols = self.protocols_of_network_for_tld(
+            network_id,
+            Tld::from_domain_name(name_with_tld).unwrap(),
+            maybe_filter,
+        )?;
+
+        let mut results = Vec::new();
+        for deployed_protocol in protocols {
+            let empty_label_hash = match &deployed_protocol.protocol.info.protocol_specific {
+                ProtocolSpecific::EnsLike(ens_like) => ens_like.empty_label_hash,
+                ProtocolSpecific::D3Connect(d3_connect) => d3_connect.empty_label_hash,
+            };
+
+            let domain_name = DomainName::new(name_with_tld, empty_label_hash)?;
+            results.push(DomainNameOnProtocol {
+                inner: domain_name,
+                deployed_protocol,
+            });
+        }
```

Comment:
it seems that it's possible to use `names_options_in_network_exact` instead of this, as the logic seems exactly the same (but with unwrap, however it's possible to return errors even if we don't expect them to happen)